src/lib/db/role-store.ts

```diff
@@ -0,0 +1,196 @@
+import { Logger, LogProvider } from '../logger';
+import NotFoundError from '../error/notfound-error';
+import { ICustomRole } from 'lib/types/model';
+import {
+    ICustomRoleInsert,
+    ICustomRoleUpdate,
+    IRoleStore,
+} from 'lib/types/stores/role-store';
+import { IRole, IUserRole } from 'lib/types/stores/access-store';
+
+const T = {
+    ROLE_USER: 'role_user',
+    ROLES: 'roles',
+};
+
+const COLUMNS = ['id', 'name', 'description', 'type'];
+
+interface IRoleRow {
+    id: number;
+    name: string;
+    description: string;
+    type: string;
+}
+
+export default class RoleStore implements IRoleStore {
+    private logger: Logger;
+
+    private eventBus: EventEmitter;
+
+    private db: Knex;
+
+    constructor(db: Knex, eventBus: EventEmitter, getLogger: LogProvider) {
+        this.db = db;
+        this.eventBus = eventBus;
+        this.logger = getLogger('lib/db/role-store.ts');
+    }
+
+    async getAll(): Promise<ICustomRole[]> {
+        const rows = await this.db
+            .select(COLUMNS)
+            .from(T.ROLES)
+            .orderBy('name', 'asc');
+
+        return rows.map(this.mapRow);
+    }
+
+    async create(role: ICustomRoleInsert): Promise<ICustomRole> {
+        const row = await this.db(T.ROLES)
+            .insert({
+                name: role.name,
+                description: role.description,
+                type: role.roleType,
+            })
+            .returning('*');
+        return this.mapRow(row[0]);
+    }
+
+    async delete(id: number): Promise<void> {
+        return this.db(T.ROLES).where({ id }).del();
+    }
+
+    async get(id: number): Promise<ICustomRole> {
+        const rows = await this.db.select(COLUMNS).from(T.ROLES).where({ id });
+        if (rows.length === 0) {
+            throw new NotFoundError(`Could not find role with id: ${id}`);
+        }
+        return this.mapRow(rows[0]);
+    }
+
+    async update(role: ICustomRoleUpdate): Promise<ICustomRole> {
+        const rows = await this.db(T.ROLES)
+            .where({
+                id: role.id,
+            })
+            .update({
+                id: role.id,
+                name: role.name,
+                description: role.description,
+            })
+            .returning('*');
+        return this.mapRow(rows[0]);
+    }
+
+    async exists(id: number): Promise<boolean> {
+        const result = await this.db.raw(
+            `SELECT EXISTS (SELECT 1 FROM ${T.ROLES} WHERE id = ?) AS present`,
+            [id],
+        );
+        const { present } = result.rows[0];
+        return present;
+    }
+
+    async nameInUse(name: string, existingId?: number): Promise<boolean> {
+        let query = this.db(T.ROLES).where({ name }).returning('id');
+        if (existingId) {
+            query = query.andWhereNot({ id: existingId });
+        }
+        const result = await query;
+        return result.length > 0;
+    }
+
+    async roleExists(name: string): Promise<boolean> {
```

Comment:
Isn't this the same as the function above nameInUse?

Reply:
Behavior isn't identical but it's no longer in use so I've just removed it, nice catch